### PR TITLE
groupby refactoring

### PR DESCRIPTION
**Pull request by czpmango**
This PR adds the following features: 
 - distinct agg (eg. `avg(distinct age)`)
 - Nested calls to AGG functions and other expressions (eg.`go+groupby+yield count(abs(age))+1`)
 - implicit groupby (eg.`go+|+yield distinct $-.age,count($-.name) where $-.age>40`)
 
*Depend on [https://github.com/vesoft-inc/nebula-common/pull/354](https://github.com/vesoft-inc/nebula-common/pull/354)*

### CODE REVIEW

File: src/validator/test/QueryValidatorTest.cpp [hunk 1]
```diff
@@ -1086,8 +1086,9 @@ TEST_F(QueryValidatorTest, GoInvalid) {
         EXPECT_FALSE(checkResult(query));
     }
     {
+        // yield agg without groupBy is not supported
         std::string query = "GO FROM \"2\" OVER like YIELD COUNT(123);";
-        EXPECT_FALSE(checkResult(query));
+        EXPECT_TRUE(checkResult(query));
```
Comment (CPWstatic) [inline src/validator/test/QueryValidatorTest.cpp:1091]: EXPECT_FALSE
```
GO FROM \"2\" OVER like WHERE COUNT(123) > 0 YIELD 123;
```

File: src/validator/test/YieldValidatorTest.cpp [hunk 2]
```diff
@@ -320,7 +320,8 @@ TEST_F(YieldValidatorTest, YieldPipe) {
 #endif
     {
         auto query = go + "| YIELD DISTINCT 1 + $-.* AS e WHERE $-.start > 2005";
-        EXPECT_FALSE(checkResult(query));
+        // + expr eval error?  (czp)
+        EXPECT_TRUE(checkResult(query));
```
Comment (CPWstatic) [inline src/validator/test/YieldValidatorTest.cpp:324]: EXPECT_FALSE

File: src/visitor/DeduceTypeVisitor.cpp [hunk 3]
```diff
@@ -450,6 +510,7 @@ void DeduceTypeVisitor::visit(InputPropertyExpression *expr) {
 
 void DeduceTypeVisitor::visit(VariablePropertyExpression *expr) {
     auto *var = expr->sym();
+    if (*var == "*") return;
```
Comment (CPWstatic) [inline src/visitor/DeduceTypeVisitor.cpp:513]: ?

File: tests/query/v1/test_groupby.py [hunk 4]
```diff
@@ -93,11 +99,11 @@ def test_group_by(self):
...
-                ["Magic", 2000, 2000, 2004, 2004.0, 0.0, 1],
-                ["Pistons", 2015, 2015, 2017, 2017.0, 0.0, 1],
-                ["Raptors", 1997, 1997, 2000, 2000.0, 0.0, 1],
-                ["Rockets", 2004, 2004, 2010, 2010.0, 0.0, 1],
+                ["Celtics", 2017, 2017, 2019, 2019, 0.0, 1],
+                ["Magic", 2000, 2000, 2004, 2004, 0.0, 1],
+                ["Pistons", 2015, 2015, 2017, 2017, 0.0, 1],
+                ["Raptors", 1997, 1997, 2000, 2000, 0.0, 1],
+                ["Rockets", 2004, 2004, 2010, 2010, 0.0, 1],
                 ["Spurs", 2013, 2013, 2013, 2014.0, 1.0, 2]
```
Comment (CPWstatic) [inline tests/query/v1/test_groupby.py:107]: Why not all double?

File: tests/query/v1/test_yield.py [hunk 5]
```diff
@@ -399,18 +399,19 @@ def test_calculate_overflow(self):
...
-        # self.check_column_names(resp, columns)
-        # expect_result = [[2, 2]]
-        # self.check_result(resp, expect_result)
+        query = '''YIELD 1+COUNT(*), 1+1'''
+        resp = self.execute(query)
+        self.check_resp_succeeded(resp)
+        columns = ["(1+COUNT(*))", "(1+1)"]
+        self.check_column_names(resp, columns)
+        expect_result = [[2, 2]]
+        self.check_result(resp, expect_result)
```
Comment (CPWstatic) [inline tests/query/v1/test_yield.py:414]: Need more cases.

File: src/validator/Validator.cpp [hunk 6]
```diff
@@ -71,8 +71,19 @@ std::unique_ptr<Validator> Validator::makeValidator(Sentence* sentence, QueryCon
...
+            auto yieldSentence = static_cast<YieldSentence*>(sentence);
+            if (yieldSentence->hasAgg()) {
+                auto* where = yieldSentence->where() ?
+                    yieldSentence->where()->clone().release() : nullptr;
+                auto* groupSentence = context->objPool()->add(
+                    new GroupBySentence(
+                    yieldSentence->yield()->clone().release(),
+                    where, nullptr));
+                return std::make_unique<GroupByValidator>(groupSentence, context);
+            }
```
Comment (CPWstatic) [inline src/validator/Validator.cpp:84]: Please move the codes to YieldValidator as we discussed offline.

File: src/validator/GroupByValidator.cpp [hunk 7]
```diff
@@ -34,58 +68,85 @@ Status GroupByValidator::validateYield(const YieldClause *yieldClause) {
         return Status::SemanticError("Yield cols is Empty");
     }
 
+    projCols_.reset(new YieldColumns);
```
[the remark this answers is not in the document]
Comment (czpmango) [reply src/validator/GroupByValidator.cpp:71]: Already fixed, thx.

File: src/validator/Validator.cpp (hunk 6 above)
Comment (czpmango) [reply src/validator/Validator.cpp:84]: Done.

File: src/validator/test/QueryValidatorTest.cpp (hunk 1 above)
Comment (czpmango) [reply src/validator/test/QueryValidatorTest.cpp:1091]: Already fixed.

File: src/validator/GoValidator.cpp [hunk 8]
```diff
@@ -108,11 +108,6 @@ Status GoValidator::validateYield(YieldClause* yield) {
                 ExpressionUtils::rewriteLabelAttribute<EdgePropertyExpression>(col->expr());
             }
 
-            if (!col->getAggFunName().empty()) {
-                return Status::SemanticError(
-                    "`%s', not support aggregate function in go sentence.",
-                    col->toString().c_str());
-            }
```
[the remark this answers is not in the document]
Comment (czpmango) [reply src/validator/GoValidator.cpp:115]: Fixed

File: src/validator/GroupByValidator.cpp [hunk 9]
```diff
@@ -34,58 +68,85 @@ Status GroupByValidator::validateYield(const YieldClause *yieldClause) {
...
-            if (iter->second != AggFun::Function::kCount && col->expr()->toString() == "*") {
-                return Status::SemanticError("`%s` invaild, * valid in count.",
-                                             col->toString().c_str());
+            if (aggs.size() == 1) {
+                auto col_copy = col->expr()->clone();
+                // set aggExpr
+                col->setExpr(aggs[0]->clone().release());
+                auto agg_col = col->toString();
+                // rewrite to VariablePropertyExpression
+                RewriteAggExprVisitor rewriteAggVisitor(new std::string(),
```
[the remark this answers is not in the document]
Comment (czpmango) [reply src/validator/GroupByValidator.cpp:91]: It's been taken over by unique_ptr.

File: src/validator/GroupByValidator.cpp [hunk 10]
```diff
@@ -34,58 +68,85 @@ Status GroupByValidator::validateYield(const YieldClause *yieldClause) {
...
 
+    projCols_.reset(new YieldColumns);
     for (auto* col : columns) {
-        auto fun = col->getAggFunName();
-        if (!fun.empty()) {
-            auto iter = AggFun::nameIdMap_.find(fun);
-            if (iter == AggFun::nameIdMap_.end()) {
-                return Status::SemanticError("Unkown aggregate function `%s`", fun.c_str());
+        auto col_name = deduceColName(col);
+        auto rewrited = 0;
```
[the remark this answers is not in the document]
Comment (czpmango) [reply src/validator/GroupByValidator.cpp]: ACK.

File: src/validator/GroupByValidator.cpp [hunk 11]
```diff
@@ -34,58 +68,85 @@ Status GroupByValidator::validateYield(const YieldClause *yieldClause) {
...
+            if (aggs.size() > 1) {
+                return Status::SemanticError("Agg function nesting is not allowed");
             }
-            if (iter->second != AggFun::Function::kCount && col->expr()->toString() == "*") {
-                return Status::SemanticError("`%s` invaild, * valid in count.",
-                                             col->toString().c_str());
+            if (aggs.size() == 1) {
+                auto col_copy = col->expr()->clone();
+                // set aggExpr
+                col->setExpr(aggs[0]->clone().release());
```
[the remark this answers is not in the document]
Comment (czpmango) [reply src/validator/GroupByValidator.cpp:88]: Got it.

File: src/validator/test/YieldValidatorTest.cpp (hunk 2 above)
Comment (czpmango) [reply src/validator/test/YieldValidatorTest.cpp]: Fixed.

File: tests/query/v1/test_groupby.py (hunk 4 above)
Comment (czpmango) [reply tests/query/v1/test_groupby.py]: Already fixed.

File: src/visitor/FoldConstantExprVisitor.cpp [hunk 12]
```diff
@@ -115,6 +115,11 @@ void FoldConstantExprVisitor::visit(FunctionCallExpression *expr) {
     canBeFolded_ = canBeFolded;
 }
 
+void FoldConstantExprVisitor::visit(AggregateExpression *expr) {
+    // TODO : impl AggExpr foldConstantExprVisitor
```
Comment (CPWstatic) [inline src/visitor/FoldConstantExprVisitor.cpp:119]: We could do folding for the args expr of agg.

**Review by CPWstatic — COMMENTED** [review]
Excellent work!

File: src/validator/Validator.h [hunk 13]
```diff
@@ -136,14 +136,17 @@ class Validator {
 
     Status invalidLabelIdentifiers(const Expression* expr) const;
 
+public:
+    // The input columns and output columns of a sentence.
+    ColsDef                         outputs_;
+    ColsDef                         inputs_;
+    ExpressionProps                 exprProps_;
```
[the remark this answers is not in the document]
Comment (czpmango) [reply src/validator/Validator.h:143]: Got it.

File: src/visitor/FoldConstantExprVisitor.cpp (hunk 12 above)
Comment (czpmango) [reply src/visitor/FoldConstantExprVisitor.cpp:119]: Already add.

File: src/validator/test/YieldValidatorTest.cpp [hunk 14]
```diff
@@ -319,7 +319,7 @@ TEST_F(YieldValidatorTest, YieldPipe) {
     }
 #endif
     {
-        auto query = go + "| YIELD DISTINCT 1 + $-.* AS e WHERE $-.start > 2005";
+        auto query = go + "| YIELD DISTINCT 1 + $-.* AS e WHERE $-.start > 2005";\
```
[the remark this answers is not in the document]
Comment (czpmango) [reply src/validator/test/YieldValidatorTest.cpp:322]: My mistake.

File: src/parser/parser.yy [hunk 15]
```diff
@@ -785,6 +789,23 @@ function_call_expression
     }
     ;
 
+aggregate_expression
+    : agg_function L_PAREN expression R_PAREN {
```
[the remark this answers is not in the document]
Comment (czpmango) [reply src/parser/parser.yy:840]: Conflict with FunctionCallExpression parser rule. 
The matching priority of agg_function is higher than the LABEL. Currently, if agg function is added, parser must be modified.

**Review by CPWstatic — DISMISSED** [review]
LGTM.

**Review by CPWstatic — DISMISSED** [review]

**Review by CPWstatic — APPROVED** [review]

File: src/parser/Sentence.h [hunk 16]
```diff
@@ -132,6 +132,10 @@ class Sentence {
         return kind_;
     }
 
+    void setKind(Kind kind) {
```
Comment (laura-ding) [inline src/parser/Sentence.h:135]: Why need add this?

File: src/parser/TraverseSentences.h [hunk 17]
```diff
@@ -563,14 +564,23 @@ class YieldSentence final : public Sentence {
...
+        yieldClause_.reset(yield);
+        havingClause_.reset(having);
+        groupClause_.reset(groupby);
     }
 
     void setGroupClause(GroupClause *clause) {
         groupClause_.reset(clause);
     }
 
+    void setHavingClause(WhereClause *clause) {
```
Comment (laura-ding) [inline src/parser/TraverseSentences.h:580]: where use it? and what's the meaning?

File: src/validator/GroupByValidator.cpp [hunk 18]
```diff
@@ -116,10 +123,131 @@ Status GroupByValidator::validateGroup(const GroupClause *groupClause) {
...
+Status GroupByValidator::checkAggExpr(AggregateExpression* aggExpr) {
+    auto func = aggExpr->name();
+    if (!func) {
+        return Status::SemanticError("`%s` aggregate function not set.",
+                                     aggExpr->toString().c_str());
+    }
+
+    auto iter = AggregateExpression::NAME_ID_MAP.find(func->c_str());
+    if (iter == AggregateExpression::NAME_ID_MAP.end()) {
+        return Status::SemanticError("Unkown aggregate function `%s`", func->c_str());
```
Comment (laura-ding) [inline src/validator/GroupByValidator.cpp:223]: unknown

File: src/validator/GroupByValidator.cpp [hunk 19]
```diff
@@ -116,10 +123,131 @@ Status GroupByValidator::validateGroup(const GroupClause *groupClause) {
...
+    }
+
+    auto* aggArg = aggExpr->arg();
+    if (graph::ExpressionUtils::findAny(aggArg,
+                                        {Expression::Kind::kAggregate})) {
+        return Status::SemanticError("Agg function nesting is not allowed: `%s`",
```
Comment (laura-ding) [inline src/validator/GroupByValidator.cpp:229]: modify the `Agg` to the whole word when as error msg.

File: tests/tck/features/agg/Agg.feature [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+    Then a SemanticError should be raised at runtime: GroupBy item `$-.dst` must be in Yield list
+    When executing query:
+      """
+      GO FROM "Tim Duncan" OVER like YIELD like._dst AS dst, $$.player.age AS age
+      | GROUP BY $-.dst,$-.x YIELD avg(distinct $-.age) AS age
+      """
+    Then a SemanticError should be raised at runtime:  `$-.x', not exist prop `x'
+    When executing query:
+      """
+      GO FROM "Tim Duncan" OVER like YIELD like._dst AS dst, $$.player.age AS age
```
Comment (laura-ding) [inline tests/tck/features/agg/Agg.feature:156]: Can you add test about this, expect execute ok
```
 GO FROM "Tim Duncan" OVER like YIELD like._dst AS dst, $$.player.age AS age
      | GROUP BY $-.age+1 YIELD $-.age+1, avg(distinct $-.age) AS age
```

File: tests/tck/features/go/GO.feature [hunk 21]
```diff
@@ -1609,12 +1609,11 @@ Feature: Go Sentence
...
     When executing query:
       """
-      GO 1 TO 2 STEPS FROM "Tim Duncan" OVER like WHERE like._dst != "YAO MING" YIELD like._dst AS vid
-      | GROUP BY $-.vid YIELD 1 AS id | GROUP BY $-.id YIELD COUNT($-.id);
+      GO 1 TO 2 STEPS FROM "Tim Duncan" OVER like WHERE like._dst != "YAO MING" YIELD like._dst AS vid | GROUP BY $-.vid YIELD $-.vid, 1 AS id | GROUP BY $-.id YIELD $-.id, COUNT($-.id);
       """
     Then the result should be, in any order, with relax comparison:
-      | COUNT($-.id) |
-      | 4            |
+      | $-.id | COUNT($-.id) |
```
Comment (laura-ding) [inline tests/tck/features/go/GO.feature:1615]: `GROUP BY $-.id YIELD  COUNT($-.id);` it the the correct representation. why need add a col `$-.id`

File: src/executor/query/AggregateExecutor.cpp [hunk 22]
```diff
@@ -35,19 +35,22 @@ folly::Future<Status> AggregateExecutor::execute() {
...
-            for (size_t i = 0; i < groupItems.size(); ++i) {
-                auto& v = groupItems[i].expr->eval(ctx(iter.get()));
-                it->second[i]->apply(v);
+        }
+
+        for (size_t i = 0; i < groupItems.size(); ++i) {
+            auto* item = groupItems[i];
+            if (item->kind() == Expression::Kind::kAggregate) {
+                static_cast<AggregateExpression*>(item)->setAggData(result[list][i].get());
+                auto eval = item->eval(ctx(iter.get()));
```
[the remark this answers is not in the document]
Comment (czpmango) [reply src/executor/query/AggregateExecutor.cpp:51]: Fixed.

File: tests/tck/features/go/GO.IntVid.feature [hunk 23]
```diff
@@ -1588,11 +1588,11 @@ Feature: IntegerVid Go  Sentence
     When executing query:
       """
       GO 1 TO 2 STEPS FROM hash("Tim Duncan") OVER like WHERE like._dst != hash("YAO MING") YIELD like._dst AS vid
-      | GROUP BY $-.vid YIELD 1 AS id | GROUP BY $-.id YIELD COUNT($-.id);
+      | GROUP BY $-.vid YIELD $-.vid, 1 AS id | GROUP BY $-.id YIELD $-.id, COUNT($-.id);
```
[the remark this answers is not in the document]
Comment (czpmango) [reply tests/tck/features/go/GO.IntVid.feature:1591]: Supported: 
```
 - group by $-.id yield $-.id ,avg($-.age)
 - yield $-.id, avg($-.age)
 - yield avg($-.age)
```
Unsupported:
```
 - group by $-.id yield avg($-.age)
```
The items of group by must appear in the yield list.

File: src/parser/Sentence.h (hunk 16 above)
Comment (czpmango) [reply src/parser/Sentence.h:135]: Not used, I'll delete it.

File: src/parser/TraverseSentences.h (hunk 17 above)
Comment (czpmango) [reply src/parser/TraverseSentences.h:580]: Not used yet, but we'll support Having clause later.

File: src/validator/GroupByValidator.cpp (hunk 19 above)
Comment (czpmango) [reply src/validator/GroupByValidator.cpp:229]: Got it.

File: src/parser/TraverseSentences.h [hunk 24]
```diff
@@ -554,6 +554,7 @@ class YieldSentence final : public Sentence {
         return yieldClause_.get();
     }
 
+    bool hasAgg();
```
[the remark this answers is not in the document]
Comment (czpmango) [reply src/parser/TraverseSentences.h:557]: ok

File: src/validator/GoValidator.cpp [hunk 25]
```diff
@@ -107,10 +111,9 @@ Status GoValidator::validateYield(YieldClause* yield) {
             } else {
                 ExpressionUtils::rewriteLabelAttribute<EdgePropertyExpression>(col->expr());
             }
-
-            if (!col->getAggFunName().empty()) {
+            if (graph::ExpressionUtils::findAny(col->expr(), {Expression::Kind::kAggregate})) {
                 return Status::SemanticError(
-                    "`%s', not support aggregate function in go sentence.",
+                    "`%s`, not support aggregate function in go sentence.",
```
[the remark this answers is not in the document]
Comment (czpmango) [reply src/validator/GoValidator.cpp:116]: Got it.

File: src/validator/GroupByValidator.cpp [hunk 26]
```diff
@@ -116,10 +123,131 @@ Status GroupByValidator::validateGroup(const GroupClause *groupClause) {
...
+        return Status::SemanticError("Not support both input and variable in GroupBy sentence.");
+    }
+    if (!exprProps_.varProps().empty() && exprProps_.varProps().size() > 1) {
+        return Status::SemanticError("Only one variable allowed to use.");
+    }
+
+    if (groupKeys_.empty()) {
+        groupKeys_ = yieldCols_;
+    } else {
+        std::unordered_set<Expression*> groupSet(begin(groupKeys_), end(groupKeys_));
```
[the remark this answers is not in the document]
Comment (czpmango) [reply src/validator/GroupByValidator.cpp:154]: Equivalent, already Modified.

File: src/validator/Validator.h [hunk 27]
```diff
@@ -62,10 +58,50 @@ class Validator {
         return outputs_;
     }
 
+    void setOutputCols(ColsDef&& outputCols) {
+        outputs_ = std::move(outputCols);
+    }
+
+    void setOutputCols(ColsDef& outputCols) {
```
[the remark this answers is not in the document]
Comment (czpmango) [reply src/validator/Validator.h:65]: ok.

File: tests/tck/features/go/GO.IntVid.feature (hunk 23 above)
[the remark this answers is not in the document]
Comment (CPWstatic) [reply tests/tck/features/go/GO.IntVid.feature:1591]: `GROUP BY $-.vid YIELD 1 AS id` Why it is not supported?

File: tests/tck/features/go/GO.feature (hunk 21 above)
Comment (czpmango) [reply tests/tck/features/go/GO.feature:1615]: We support yield sentence that only has aggregate columns, but the items of group by must appear in the yield list.
Similar to the discussion here: `https://github.com/vesoft-inc/nebula-graph/pull/503#discussion_r553698798`

File: tests/tck/features/go/GO.IntVid.feature (hunk 23 above)
[the remark this answers is not in the document]
Comment (czpmango) [reply tests/tck/features/go/GO.IntVid.feature:1591]: > `GROUP BY $-.vid YIELD 1 AS id` Why it is not supported?

Now it's going to report an error:` SemanticError: GroupBy item `$-.dst` must be in Yield list`.
But, Do we plan to support this : `GROUP BY 1 YIELD 1 AS id`

File: src/executor/test/AggregateTest.cpp [hunk 28]
```diff
@@ -105,12 +105,14 @@ struct RowCmp {
 
 #define TEST_AGG_2(FUN, COL, DISTINCT)                                      \
     std::vector<Expression*> groupKeys;                                     \
-    std::vector<Aggregate::GroupItem> groupItems;                           \
-    auto expr =                                                             \
+    std::vector<Expression*> groupItems;                           \
+    auto expr1 =                                                             \
+        std::make_unique<InputPropertyExpression>(new std::string("col2")); \
+    auto expr2 =                                                             \
```
[the remark this answers is not in the document]
Comment (czpmango) [reply src/executor/test/AggregateTest.cpp:111]: ok

File: tests/tck/features/go/GO.feature (hunk 21 above)
Comment (laura-ding) [reply tests/tck/features/go/GO.feature:1615]: `GROUP BY $-.id YIELD COUNT($-.id);` 
the group item `$-.id ` is in yield clause, the `COUNT($-.id);`  means include it.

File: src/executor/test/AggregateTest.cpp [hunk 29]
```diff
@@ -85,11 +85,11 @@ struct RowCmp {
 
 #define TEST_AGG_1(FUN, COL, DISTINCT)                                      \
```
[the remark this answers is not in the document]
Comment (czpmango) [reply src/executor/test/AggregateTest.cpp:86]: Line 96 used it.

File: tests/tck/features/go/GO.feature (hunk 21 above)
Comment (czpmango) [reply tests/tck/features/go/GO.feature:1615]: Already fixed. Need not col $-.id.

File: tests/tck/features/agg/Agg.feature (hunk 20 above)
Comment (czpmango) [reply tests/tck/features/agg/Agg.feature:157]: Already added.

File: tests/query/v1/test_groupby.py [hunk 30]
```diff
@@ -0,0 +1,424 @@
...
+from tests.common.nebula_test_suite import NebulaTestSuite, T_NULL_BAD_TYPE
+
+
+class TestGroupBy(NebulaTestSuite):
+    @classmethod
+    def prepare(self):
+        self.load_data()
+
+    def test_syntax_error(self):
+        # Use groupby without input
```
[the remark this answers is not in the document]
Comment (czpmango) [reply tests/query/v1/test_groupby.py]: Already moved.

**Review by CPWstatic — DISMISSED** [review]

File: src/validator/YieldValidator.cpp [hunk 31]
```diff
@@ -44,39 +51,21 @@ Status YieldValidator::validateImpl() {
...
-    if (exprProps_.inputProps().empty() && exprProps_.varProps().empty() && inputVarName_.empty()) {
+    if (!groupByValidator_ && exprProps_.inputProps().empty()
+        && exprProps_.varProps().empty() && inputVarName_.empty()) {
         // generate constant expression result into querycontext
         genConstantExprValues();
     }
 
     if (!exprProps_.varProps().empty() && !userDefinedVarNameList_.empty()) {
+        // TODO: Multiple userDefinedVars not support yet, may be supported later
+        DCHECK_EQ(userDefinedVarNameList_.size(), 1);
```
Comment (laura-ding) [inline src/validator/YieldValidator.cpp:62]: here return error directly,  not use DCHECK.

**Review by laura-ding — APPROVED** [review]

File: src/validator/YieldValidator.cpp (hunk 31 above)
Comment (czpmango) [reply src/validator/YieldValidator.cpp]: Modified~

**Review by laura-ding — APPROVED** [review]

**Review by yixinglu — APPROVED** [review]

**Pull request merged**